### ISSUE TITLE
`AttributeDict` typing and constructor updates

### ISSUE THREAD

**Issue by pacrob** [ethereum/web3.py#3105]
`AttributeDict` and it's parent classes use `Mapping[TKey, TValue]` for typing, but in practice they often have multiple types in a single structure, at least in the values (strings, ints, hexbytes, etc.)

Consider changing the type, either by being more specific, or opening it to `Any`, as a user could theoretically turn any mapping into an `AttributeDict`.

Also, since an `AttributeDict` is expected to be a hashable datastructure, consider doing recursive tupleization/checking for unhashable types at the time of creation, instead of only when trying to hash it.

**Comment by fselmo** [comment]
I took a quick peek at this. The way it's currently set up allows the user to define what they want it to be. For example,

```py
# this still works with `Any`
some_arg: AttributeDict[str, Any]

# but if we want to narrow the typing more, we can
some_arg: AttributeDict[bytes, bytes]
```

Unless I'm missing something, I think this is good the way it is. Thoughts?

**Comment by fselmo** [comment]
> Also, since an AttributeDict is expected to be a hashable datastructure, consider doing recursive tupleization/checking for unhashable types at the time of creation, instead of only when trying to hash it.

I guess I didn't address this part but I feel like we shouldn't change the types that aren't assigned internally, especially if it is indeed declared at the type-hinting level. I'd rather only use it when hashing, as we currently do.

**Comment by pacrob** [comment]
Yep, no issue leaving as-is. I don't think I understood the creation process properly when I opened the issue.
